pkg/sigstorebundle/store.go

```
@@ -0,0 +1,279 @@
+	}
+
+	if len(payload) == 0 {
+		return nil, fmt.Errorf("decoded payload is empty")
+	}
+
+	env := &dsse.Envelope{
+		Payload:     payload,
+		PayloadType: bundle.DsseEnvelope.PayloadType,
+	}
+
+	if len(bundle.DsseEnvelope.Signatures) == 0 {
+		return nil, fmt.Errorf("bundle has no signatures")
+	}
+
+	// Check signature count limit to prevent resource exhaustion
+	if len(bundle.DsseEnvelope.Signatures) > bundleLimits.MaxSignaturesPerBundle {
+		return nil, fmt.Errorf("bundle has %d signatures, exceeds maximum allowed (%d)", len(bundle.DsseEnvelope.Signatures), bundleLimits.MaxSignaturesPerBundle)
+	}
+
+	// Map signatures with VerificationMaterial
+	for idx, bundleSig := range bundle.DsseEnvelope.Signatures {
+		if bundleSig.Sig == "" {
+			return nil, fmt.Errorf("signature %d missing sig field", idx)
+		}
+
+		sig, err := base64.StdEncoding.DecodeString(bundleSig.Sig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode signature %d: %w", idx, err)
+		}
+
+		if len(sig) == 0 {
+			return nil, fmt.Errorf("decoded signature %d is empty", idx)
+		}
+
+		witnessSig := dsse.Signature{
+			KeyID:     bundleSig.KeyID,
+			Signature: sig,
+		}
+
+		// Map VerificationMaterial to signature fields
+		if bundle.VerificationMaterial != nil {
+			mapVerificationMaterial(&witnessSig, bundle.VerificationMaterial)
+		}
+
+		env.Signatures = append(env.Signatures, witnessSig)
+	}
+
+	return env, nil
+}
+
+// mapVerificationMaterial maps Sigstore VerificationMaterial to go-witness signature fields
+func mapVerificationMaterial(sig *dsse.Signature, vm *VerificationMaterial) {
+	if sig == nil || vm == nil {
+		return
+	}
+
+	// Certificate chain
+	if vm.X509CertificateChain != nil && len(vm.X509CertificateChain.Certificates) > 0 {
+		// First cert is leaf
+		if vm.X509CertificateChain.Certificates[0].RawBytes != "" {
+			if cert, err := base64.StdEncoding.DecodeString(vm.X509CertificateChain.Certificates[0].RawBytes); err == nil {
```

looks like we are silently dropping when base64 decoding fails. An attacker could corrupt the data mid flight, causing data to be stored without the certificate chains. We should fail the upload when cert chains or timestamps are corrupted.